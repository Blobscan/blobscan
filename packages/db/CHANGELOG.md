# @blobscan/db

## 0.1.0

### Minor Changes

- [#235](https://github.com/Blobscan/blobscan/pull/235) [`71887f8`](https://github.com/Blobscan/blobscan/commit/71887f8dc09b45b0cb0748c0d6e8ddce2662f34d) Thanks [@PJColombo](https://github.com/PJColombo)! - - Add Support For Block Reorganizations:

  - Prisma:
    - Create `TransactionFork` model.
    - Utilize the block's hash as a foreign key instead of the block number.
  - API:

    - Introduce an indexer's TRPC procedure/endpoint to manage reorganizations.
    - Enable retrieval of either normal or reorganized blocks in block getter procedures.

  - Update Blockchain Synced State by Monitoring Lower and Upper Synced Slots:

    - Prisma:
      - Add `last_lower_synced_slot` and `last_upper_synced_slot` fields to `BlockchainSyncState` model.
    - API:
      - Move the blockchain synced state updater procedure from the indexer's router to its dedicated router.
      - Overhaul blockchain synced state procedures for updating and retrieving the entire entity, rather than making granular updates.

  - Make the Last Finalized Block Optional.

- [#252](https://github.com/Blobscan/blobscan/pull/252) [`534545c`](https://github.com/Blobscan/blobscan/commit/534545c321e716d24f2d73d89660271610189a8a) Thanks [@PJColombo](https://github.com/PJColombo)! - Eliminated the need for the Beacon API endpoint in stats calculation by tracking the last finalized block in the database and utilizing it internally.

- [#259](https://github.com/Blobscan/blobscan/pull/259) [`5ed5186`](https://github.com/Blobscan/blobscan/commit/5ed51867d5b01b0572bfa69f3211a5b5bfaf254e) Thanks [@0xGabi](https://github.com/0xGabi)! - Added kzg proof field to Blob table

### Patch Changes

- [#247](https://github.com/Blobscan/blobscan/pull/247) [`2fb02b0`](https://github.com/Blobscan/blobscan/commit/2fb02b0268e1fcafc10abefb079d822845392d73) Thanks [@0xGabi](https://github.com/0xGabi)! - Set up changeset

- [#271](https://github.com/Blobscan/blobscan/pull/271) [`acf8dff`](https://github.com/Blobscan/blobscan/commit/acf8dff07d62a33d5dfa3789fd1f4e2aa3e968a3) Thanks [@0xGabi](https://github.com/0xGabi)! - Updated blob kzg proof field to non unique

- [#252](https://github.com/Blobscan/blobscan/pull/252) [`263d86f`](https://github.com/Blobscan/blobscan/commit/263d86ff5e35f7cf9d5bf18f6b745f8dd6249e62) Thanks [@PJColombo](https://github.com/PJColombo)! - Improved date normalization error message

- [#252](https://github.com/Blobscan/blobscan/pull/252) [`cb732e7`](https://github.com/Blobscan/blobscan/commit/cb732e7febcbc05bdec2221fec1213bcb8172717) Thanks [@PJColombo](https://github.com/PJColombo)! - Resolved Date normalization issue

- Updated dependencies [[`2bce443`](https://github.com/Blobscan/blobscan/commit/2bce443401b1875df40298ebd957f86a92539397), [`2fb02b0`](https://github.com/Blobscan/blobscan/commit/2fb02b0268e1fcafc10abefb079d822845392d73)]:
  - @blobscan/logger@0.0.2
  - @blobscan/open-telemetry@0.0.2
  - @blobscan/dayjs@0.0.2
