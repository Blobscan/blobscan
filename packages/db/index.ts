export * from "@prisma/client";
export {
  PrismaClientKnownRequestError,
  PrismaClientUnknownRequestError,
  PrismaClientValidationError,
} from "@prisma/client/runtime/library";
export type {
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@prisma/client/runtime/library";

export * from "./prisma";
