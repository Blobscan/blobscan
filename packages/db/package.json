{
  "name": "@blobscan/db",
  "version": "0.0.1",
  "private": true,
  "main": "./index.ts",
  "types": "./index.ts",
  "license": "MIT",
  "scripts": {
    "clean": "rm -rf .turbo node_modules",
    "type-check": "tsc --noEmit",
    "db:generate": "pnpm with-env prisma generate",
    "db:push": "pnpm with-env prisma db push --skip-generate",
    "db:migrate": "pnpm with-env prisma migrate dev",
    "db:seed": "pnpm with-env prisma db seed",
    "db:reset": "pnpm with-env prisma db push --force-reset",
    "dev": "pnpm with-env prisma studio --port 5556",
    "with-env": "dotenv -e ../../.env --"
  },
  "prisma": {
    "seed": "ts-node prisma/seed.ts"
  },
  "dependencies": {
    "@blobscan/blob-storage-manager": "^0.0.1",
    "@blobscan/dayjs": "^0.0.1",
    "@prisma/client": "^5.0.0"
  },
  "devDependencies": {
    "@blobscan/eslint-config": "^0.0.1",
    "@faker-js/faker": "^8.0.2",
    "js-sha256": "^0.9.0",
    "prisma": "^5.0.0"
  },
  "eslintConfig": {
    "root": true,
    "extends": [
      "@blobscan/eslint-config/base"
    ]
  }
}
