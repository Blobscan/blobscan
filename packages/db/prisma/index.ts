import { PrismaClient } from "@prisma/client";

import { baseExtension, statsExtension } from "./extensions";

const globalForPrisma = globalThis as unknown as { prisma: PrismaClient };

const prisma_ =
  globalForPrisma.prisma ||
  new PrismaClient({
    log:
      process.env.NODE_ENV === "development"
        ? ["query", "error", "warn"]
        : ["error"],
  });

export const prisma = prisma_.$extends(baseExtension).$extends(statsExtension);

export type BlobscanPrismaClient = typeof prisma;

export * from "./types";
export * from "./utils/dates";

if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma_;
