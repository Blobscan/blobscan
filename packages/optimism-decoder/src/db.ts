import { getPrisma } from "@blobscan/db";

import type { OptimismDecodedData } from "./decoder";

export function saveDecodedOptimismDataToDB({
  hash,
  data,
}: {
  hash: string;
  data: OptimismDecodedData;
}) {
  return getPrisma().transaction.update({
    where: {
      hash,
      //rollup: "OPTIMISM",
    },
    data: {
      decodedFields: data,
    },
  });
}
