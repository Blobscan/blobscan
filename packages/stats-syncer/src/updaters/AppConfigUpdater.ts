import type { Redis } from "ioredis";

import { normalizeDailyDate, normalizeDate, prisma } from "@blobscan/db";
import type { PrismaPromise, RawDatePeriod } from "@blobscan/db";

import { PeriodicUpdater } from "../PeriodicUpdater";
import { formatDate, log } from "../utils";

interface DailyStatsModel {
  findFirst: (args: {
    select: { day: boolean };
    orderBy: { day: "desc" };
  }) => PrismaPromise<{ day: Date } | null>;
  populate: (datePeriod: RawDatePeriod) => PrismaPromise<number>;
}

const dailyStatsModels: Record<string, DailyStatsModel> = {
  blob: prisma.blobDailyStats,
  block: prisma.blockDailyStats,
  transaction: prisma.transactionDailyStats,
};

export class DailyStatsUpdater extends PeriodicUpdater {
  constructor(redisUriOrConnection: string | Redis) {
    const name = "daily-stats-syncer";
    super({
      name,
      redisUriOrConnection,
      updaterFn: async () => {
        const findLatestArgs: {
          select: {
            day: boolean;
          };
          orderBy: {
            day: "desc";
          };
        } = {
          select: {
            day: true,
          },
          orderBy: {
            day: "desc",
          },
        };

        const lastIndexedBlock = await prisma.block.findLatest();

        if (!lastIndexedBlock) {
          log("debug", "Skipping stats aggregation. No blocks indexed yet", {
            updater: name,
          });

          return;
        }
        const targetDate = normalizeDate(lastIndexedBlock.timestamp).subtract(
          1,
          "day"
        );
        const targetDay = normalizeDailyDate(targetDate);

        const lastDailyStatsDays = await Promise.all(
          Object.values(dailyStatsModels).map((m) =>
            m
              .findFirst(findLatestArgs)
              .then((stats) =>
                stats?.day ? normalizeDate(stats.day) : undefined
              )
          )
        );

        if (
          lastDailyStatsDays.every((lastDay) =>
            lastDay ? lastDay?.isSame(targetDay, "day") : false
          )
        ) {
          log("debug", `Skipping stats aggregation. Already up to date`, {
            updater: name,
          });

          return;
        }

        const entityToPopulatedDays = await Promise.all(
          Object.entries(dailyStatsModels).map(async ([entity, model], i) => {
            const lastDailyStatsDay = lastDailyStatsDays[i];
            const startDate = lastDailyStatsDay
              ? lastDailyStatsDay.add(1, "day")
              : undefined;

            const populatedDays = await model.populate({
              from: startDate,
              to: targetDay,
            });

            return [entity, populatedDays];
          })
        );

        const results = entityToPopulatedDays
          .map(
            ([entity, populatedDays]) =>
              `${populatedDays} ${entity} daily stats created`
          )
          .join(", ");

        log(
          "info",
          `Daily data up to day ${formatDate(
            targetDay
          )} aggregated. ${results} successfully.`,
          {
            updater: name,
          }
        );
      },
    });
  }
}
