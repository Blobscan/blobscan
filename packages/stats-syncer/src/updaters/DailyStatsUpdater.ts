import type { Redis } from "ioredis";

import dayjs from "@blobscan/dayjs";
import { prisma } from "@blobscan/db";
import type { RawDatePeriod } from "@blobscan/db";

import { PeriodicUpdater } from "../PeriodicUpdater";
import { log } from "../utils";

export class DailyStatsUpdater extends PeriodicUpdater {
  constructor(redisUriOrConnection: string | Redis) {
    const name = "daily-stats-syncer";
    super({
      name,
      redisUriOrConnection,
      updaterFn: async () => {
        const yesterday = dayjs().subtract(1, "day");
        const datePeriod: RawDatePeriod = {
          from: yesterday.startOf("day"),
          to: yesterday.endOf("day"),
        };

        await Promise.all([
          prisma.blobDailyStats.populate(datePeriod),
          prisma.blockDailyStats.populate(datePeriod),
          prisma.transactionDailyStats.populate(datePeriod),
        ]);

        log(
          "info",
          `Day ${yesterday.format(
            "YYYY-MM-DD"
          )} stats aggregated successfully.`,
          {
            updater: name,
          }
        );
      },
    });
  }
}
