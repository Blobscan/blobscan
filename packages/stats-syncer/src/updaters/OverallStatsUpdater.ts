import type { Redis } from "ioredis";

import { prisma } from "@blobscan/db";
import type { BlockNumberRange } from "@blobscan/db";

import { PeriodicUpdater } from "../PeriodicUpdater";
import { log } from "../utils";

export class OverallStatsUpdater extends PeriodicUpdater {
  constructor(redisUriOrConnection: string | Redis) {
    const name = "overall-stats-syncer";
    super({
      name,
      redisUriOrConnection,
      updaterFn: async () => {
        const { lastAggregatedBlock, lastFinalizedBlock } =
          (await prisma.blockchainSyncState.findUnique({
            select: {
              lastAggregatedBlock: true,
              lastFinalizedBlock: true,
            },
            where: {
              id: 1,
            },
          })) ?? { lastAggregatedBlock: null, lastFinalizedBlock: null };

        if (lastFinalizedBlock === null) {
          log(
            "debug",
            `Skipping overall stats aggregation. No blocks have been indexed yet`,
            {
              updater: name,
            }
          );

          return;
        }

        const blockRange: BlockNumberRange = {
          from: lastAggregatedBlock ?? 0,
          to: lastFinalizedBlock,
        };

        if (blockRange.from >= blockRange.to) {
          log(
            "debug",
            `Skipping overall stats aggregation. No new finalized blocks (last aggregated block: ${lastAggregatedBlock})`,
            {
              updater: name,
            }
          );

          return;
        }

        await prisma.$transaction([
          prisma.blockOverallStats.increment(blockRange),
          prisma.transactionOverallStats.increment(blockRange),
          prisma.blobOverallStats.increment(blockRange),
          prisma.blockchainSyncState.upsert({
            create: {
              lastAggregatedBlock: blockRange.to,
            },
            update: {
              lastAggregatedBlock: blockRange.to,
            },
            where: {
              id: 1,
            },
          }),
        ]);

        log(
          "info",
          `Blocks ${blockRange.from} to ${blockRange.to} stats aggregated successfully.`,
          {
            updater: name,
          }
        );
      },
    });
  }
}
