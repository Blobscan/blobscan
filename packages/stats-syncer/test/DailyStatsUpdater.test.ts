import { beforeEach, describe, expect, it } from "vitest";

import { prisma } from "@blobscan/db";

import { DailyStatsUpdater } from "../src/updaters/DailyStatsUpdater";

class DailyStatsUpdaterMock extends DailyStatsUpdater {
  constructor(redisUri = process.env.REDIS_URI ?? "") {
    super(redisUri);
  }

  getWorker() {
    return this.worker;
  }

  getWorkerProcessor() {
    return this.updaterFn;
  }

  getQueue() {
    return this.queue;
  }
}

describe("DailyStatsUpdater", () => {
  let dailyStatsUpdater: DailyStatsUpdaterMock;

  beforeEach(() => {
    dailyStatsUpdater = new DailyStatsUpdaterMock();

    return async () => {
      await dailyStatsUpdater.close();
    };
  });

  it("should aggregate yesterday's stats correctly", async () => {
    const workerProcessor = dailyStatsUpdater.getWorkerProcessor();

    await workerProcessor();

    const dailyStats = await Promise.all([
      prisma.blobDailyStats.findMany(),
      prisma.blockDailyStats.findMany(),
      prisma.transactionDailyStats.findMany(),
    ]);

    expect(dailyStats).toMatchSnapshot();
  });
});
