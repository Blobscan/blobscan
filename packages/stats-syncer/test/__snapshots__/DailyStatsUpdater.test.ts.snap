// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DailyStatsUpdater > should aggregate yesterday's stats correctly 1`] = `
[
  [
    {
      "avgBlobSize": 1280,
      "day": 2023-08-31T00:00:00.000Z,
      "totalBlobSize": 6400n,
      "totalBlobs": 5,
      "totalUniqueBlobs": 3,
    },
  ],
  [
    {
      "avgBlobAsCalldataFee": 5610000,
      "avgBlobFee": 121000000,
      "avgBlobGasPrice": 22,
      "day": 2023-08-31T00:00:00.000Z,
      "totalBlobAsCalldataFee": "16830000",
      "totalBlobAsCalldataGasUsed": "765000",
      "totalBlobFee": "363000000",
      "totalBlobGasUsed": "16500000",
      "totalBlocks": 3,
    },
  ],
  [
    {
      "avgMaxBlobGasFee": 100,
      "day": 2023-08-31T00:00:00.000Z,
      "totalTransactions": 3,
      "totalUniqueReceivers": 2,
      "totalUniqueSenders": 2,
    },
  ],
]
`;
